# DataScience Mini Project
- This is a mini project on data science that tells you a basics of data science. Data Science is highly demanding and highly paid career in future.

In this project, I performed basics operation on titanic dataset.

In [ ]:
import pandas as pd     # importing pandas library
titanic_df = pd.read_csv("./titanic.csv")       # load titanic dataset into titanic_df

In [3]:
titanic_df.head()       # display quick overview of dataset

survived  pclass     sex   age  sibsp  parch     fare embarked  class   
0         0       3    male  22.0      1      0   7.2500        S  Third  \
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

`The dataset contains different properties of passengers e.g, age, sex, who, fare, etc and survived feature.`

Now, in the dataset, there are many missing values that are not present in the dataset. Usually, these are represented by `NaN`, `NA`, `null` or `None`. 

### Disadvantages of Ignoring missing values:

- If there are large numbers of missing values in the dataset and we remove them, then it may be data loss. 
- If we do not handle them and perform some machine learning tasks, it may efficieny and various losses and much more.<br>
So we have to handle them. There are many methods to handle them:

## 1. Simple Imputation Methods (Mean, Median Mode):
- **Mean and Medium**: Suitable for numerical data.
- **Mode**: Suitable for categorical data.

In [4]:
# first we need to check missing values in the loaded dataset
titanic_df.isnull().sum()    # check for missing values in the dataset

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [ ]:
titanic_df.shape    # check the shape of the dataset

(891, 15)

`See in above outputs, we have total 891 rows and 15 columns and there are 177 missing values in age column and 688 missing values in deck column. If we remove these rows of missing values, then 688 rows will be reomved from 891 rows. In return, there is only few rows in dataset. So in this case, dropping rows causing data loss.`<br><br>
`In each dataset we have to examine each feature of the data. Its purpose is to determine which features are useful for us and which features are not. In the given dataset, we have to determine wheather the pessenger is survived or not on the basis of other features. Deck column has not importance in determining the survival of passenger. So we can easily remove the column from dataset.`

In [6]:
titanic_df.drop("deck", axis=1, inplace=True) # drop the 'deck' column
titanic_df.isnull().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
embark_town      2
alive            0
alone            0
dtype: int64

**There is still 177 missing values in age column if we remove it, it also causes data loss.**

In [7]:
# Mean iputation
titanic_df["age"] = titanic_df["age"].fillna(titanic_df["age"].mean())      # fill missing values in 'age' column with mean
titanic_df.isnull().sum()

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       2
class          0
who            0
adult_male     0
embark_town    2
alive          0
alone          0
dtype: int64

In 'embarked' and 'embark_town' columns, there are two missing values. 2 or 4 rows do not effect by 890 rows

In [8]:
# Median imputation

import pandas as pd     # importing pandas library
titanic_df = pd.read_csv("./titanic.csv")       # load titanic dataset into titanic_df
titanic_df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class   
0         0       3    male  22.0      1      0   7.2500        S  Third  \
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [9]:
titanic_df.isnull().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [10]:
# dropping 'deck' feature
titanic_df.drop('deck', axis=1, inplace=True)

In [12]:
titanic_df['age'] = titanic_df['age'].fillna(titanic_df['age'].median())  # fill missing values in 'Age' column with median
titanic_df.isnull().sum()

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       2
class          0
who            0
adult_male     0
embark_town    2
alive          0
alone          0
dtype: int64

In [13]:
# mode imputation
titanic_df = pd.read_csv("./titanic.csv")
titanic_df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class   
0         0       3    male  22.0      1      0   7.2500        S  Third  \
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [14]:
titanic_df.isnull().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [15]:
titanic_df.drop('deck', axis=1, inplace=True)

In [16]:
titanic_df['embarked'] = titanic_df['embarked'].fillna(titanic_df['embarked'].mode()[0])    # imputing missing values in embarked column with first mode
titanic_df['embark_town'] = titanic_df['embark_town'].fillna(titanic_df['embark_town'].mode()[0])    # imputing missing values in embark_town column with first mode
titanic_df.isnull().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         0
class            0
who              0
adult_male       0
embark_town      0
alive            0
alone            0
dtype: int64

`These are three Simple imputation methods.` <br>
`We can also impute missing values using Machine Learning algorithms`

## 2. K-Nearest Neighbours (KNN):
KNN is a machine learning algorithm that can be used for imputing missing values. It works by finding the most similar data points to the one with the missing value based on other available features. The missing value is then imputed with the mean or median of the most similar data points.<br>
You can see the brief description of this algorithm:<br>
<a href="https://scikit-learn.org/stable/modules/generated/sklearn.impute.KNNImputer.html">KNN Imputer</a>

In [18]:
from sklearn.impute import KNNImputer   # importing KNNImputer from sklearn
titanic_df = pd.read_csv("./titanic.csv")   # loading titanic dataset 

imputer = KNNImputer(n_neighbors=5)   # creating an instance of KNNImputer with 5 neighbors

#impute missing values with KNN imputer
titanic_df['age'] = imputer.fit_transform(titanic_df[['age']])

# check the number of missing values in each column
titanic_df.isnull().sum().sort_values(ascending=False)

deck           688
embarked         2
embark_town      2
survived         0
pclass           0
sex              0
age              0
sibsp            0
parch            0
fare             0
class            0
who              0
adult_male       0
alive            0
alone            0
dtype: int64

## 3. Reggression Imputation
Regression imputation uses a regression model to predict the missing values based on other variables in the dataset. It works well for both categorical and numerical data.<br>
You can see the brief description of this algorithm:<br>
<a href="https://scikit-learn.org/stable/modules/generated/sklearn.impute.IterativeImputer.html">Iterative Imputer</a>

In [20]:
titanic_df = pd.read_csv("./titanic.csv")
titanic_df.isnull().sum().sort_values(ascending=False)

deck           688
age            177
embarked         2
embark_town      2
survived         0
pclass           0
sex              0
sibsp            0
parch            0
fare             0
class            0
who              0
adult_male       0
alive            0
alone            0
dtype: int64

In [22]:
# impute missing values with regression imputer
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer

imputer = IterativeImputer(max_iter=10) # calling the IteravtiveImputer class with max_iter = 10

# imputing missing vales in age column
titanic_df['age'] = imputer.fit_transform(titanic_df[['age']])

titanic_df.isnull().sum().sort_values(ascending=False)

deck           688
embarked         2
embark_town      2
survived         0
pclass           0
sex              0
age              0
sibsp            0
parch            0
fare             0
class            0
who              0
adult_male       0
alive            0
alone            0
dtype: int64

You can also use various ML imputer from sickit learn. Its your task to learn different models from <a href="https://scikit-learn.org/">Sickit Learn</a> to master machine learning.